File: Go/2024/2024-003/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

// Function to calculate the sum of all mul(X, Y) operations
func sumMulOperations(corruptedMemory string) int {
	// Define the regex pattern for mul(X,Y)
	pattern := regexp.MustCompile(`mul\((\d{1,3}),(\d{1,3})\)`)
	matches := pattern.FindAllStringSubmatch(corruptedMemory, -1)
	total := 0

	// Iterate over all matches and calculate the sum of products
	for _, match := range matches {
		num1, _ := strconv.Atoi(match[1])
		num2, _ := strconv.Atoi(match[2])
		total += num1 * num2
	}

	return total
}

// Function to calculate the sum of enabled mul(X, Y) operations
func sumEnabledMulOperations(corruptedMemory string) int {
	// Define the regex pattern for do(), don't(), and mul(X,Y)
	pattern := regexp.MustCompile(`(do\(\)|don't\(\)|mul\((\d{1,3}),(\d{1,3})\))`)
	matches := pattern.FindAllStringSubmatch(corruptedMemory, -1)
	totalSum := 0
	mulEnabled := true // mul instructions are enabled at the start

	// Iterate over all matches
	for _, match := range matches {
		fullMatch := match[1]

		if fullMatch == "do()" {
			mulEnabled = true
		} else if fullMatch == "don't()" {
			mulEnabled = false
		} else if match[2] != "" && match[3] != "" && mulEnabled {
			// This is a mul(X, Y) instruction
			num1, _ := strconv.Atoi(match[2])
			num2, _ := strconv.Atoi(match[3])
			totalSum += num1 * num2
		}
	}

	return totalSum
}

func main() {
	// Read the corrupted memory from 'input.txt'
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error: Unable to open file input.txt")
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	corruptedMemory := ""
	for scanner.Scan() {
		corruptedMemory += scanner.Text()
	}

	// Calculate the results
	totalSumAllMulOperations := sumMulOperations(corruptedMemory)
	totalSumEnabledMulOperations := sumEnabledMulOperations(corruptedMemory)

	// Output the results
	fmt.Printf("The sum of all mul operations is: %d\n", totalSumAllMulOperations)
	fmt.Printf("The sum of enabled mul operations is: %d\n", totalSumEnabledMulOperations)
}

File: C#/2024/2024-003/main.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

class Program
{
    // Function to sum all mul(X, Y) operations
    static int SumMulOperations(string corruptedMemory)
    {
        string pattern = @"mul\((\d{1,3}),(\d{1,3})\)";
        Regex regex = new Regex(pattern);
        MatchCollection matches = regex.Matches(corruptedMemory);
        int total = 0;

        foreach (Match match in matches)
        {
            int num1 = int.Parse(match.Groups[1].Value);
            int num2 = int.Parse(match.Groups[2].Value);
            total += num1 * num2;
        }

        return total;
    }

    // Function to sum only enabled mul(X, Y) operations
    static int SumEnabledMulOperations(string corruptedMemory)
    {
        string pattern = @"(do\(\)|don't\(\)|mul\((\d{1,3}),(\d{1,3})\))";
        Regex regex = new Regex(pattern);
        MatchCollection matches = regex.Matches(corruptedMemory);
        int totalSum = 0;
        bool mulEnabled = true; // mul instructions are enabled at the start

        foreach (Match match in matches)
        {
            string fullMatch = match.Groups[1].Value;

            if (fullMatch == "do()")
            {
                mulEnabled = true;
            }
            else if (fullMatch == "don't()")
            {
                mulEnabled = false;
            }
            else if (match.Groups[2].Success && match.Groups[3].Success)
            {
                // This is a mul(X, Y) instruction
                if (mulEnabled)
                {
                    int num1 = int.Parse(match.Groups[2].Value);
                    int num2 = int.Parse(match.Groups[3].Value);
                    totalSum += num1 * num2;
                }
            }
        }

        return totalSum;
    }

    static void Main(string[] args)
    {
        // Read the corrupted memory from 'input.txt'
        string filePath = "input.txt";
        string corruptedMemory = File.ReadAllText(filePath);

        // Calculate the results
        int totalSumAllMulOperations = SumMulOperations(corruptedMemory);
        int totalSumEnabledMulOperations = SumEnabledMulOperations(corruptedMemory);

        // Output the results
        Console.WriteLine($"The sum of all mul operations is: {totalSumAllMulOperations}");
        Console.WriteLine($"The sum of enabled mul operations is: {totalSumEnabledMulOperations}");
    }
}
